include/dali/core/exec/tasking/scheduler.h

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_EXEC_TASKING_SCHEDULER_H_
+#define DALI_CORE_EXEC_TASKING_SCHEDULER_H_
+
+#include <mutex>
+#include <condition_variable>
+#include <queue>
+#include <utility>
+#include <vector>
+#include "dali/core/api_helper.h"
+#include "dali/core/exec/tasking/task.h"
+#include "dali/core/exec/tasking/sync.h"
+
+namespace dali::tasking {
+
+/** Represents a future result of a task.
+ *
+ * Like with std::future, this object can be used to wait for and obtain a result of a task.
+ * The difference is that it needs a Scheduler object which can be used to determine task's
+ * readiness.
+ * TaskFuture prolongs the life ot a task object and its result.
+ */
+class TaskFuture {
```

Comment:
A conceptual question. STL separates waiting and obtaining the future value. Do you think there is a value in that (like waiting for void futures so getting its value is pointless)?

Reply:
STL future::get also waits, so we're not that far off. Regarding `get` for void - yes, I think it still makes sense because it makes it impossible to miss exceptions.